### PR TITLE
Async Lock

### PR DESCRIPTION
**Pull request by jcdickinson**
Low allocation lock (zero if there is no contention) that can be used in async.

### CODE REVIEW

File: src/SourceCode.Clay.Threading/AsyncLock.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+            if (wait.IsCompleted)
+                return new ValueTask<AsyncLockCookie>(Wait(wait.Result));
+            else
+                return ImplAsync(wait);
+        }
+
+        private bool TryAcquire(out ValueTask<AsyncLockCookie> result, bool ping, CancellationToken cancellationToken)
+        {
+            var acquired = _semaphore.Wait(0, cancellationToken);
```
Comment (jcdickinson) [inline src/SourceCode.Clay.Threading/AsyncLock.cs]: Prevents allocation of a Task or associated state machine if the semaphore can be acquired immediately.

File: src/SourceCode.Clay.Threading/AsyncLock.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace SourceCode.Clay.Threading
+{
+    /// <summary>
+    /// Represents a lock that can operate under an asynchronous context.
+    /// </summary>
+    public sealed partial class AsyncLock : IDisposable
```
Comment (grant-d) [inline src/SourceCode.Clay.Threading/AsyncLock.cs]: Doesn't look like you're using `partial` anywhere

File: src/SourceCode.Clay.Threading/AsyncLock.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            : new AsyncLockCookie();
+
+        private ValueTask<AsyncLockCookie> WaitAsync(Task<bool> wait)
+        {
+            async ValueTask<AsyncLockCookie> ImplAsync(Task<bool> task)
+                => Wait(await task);
+
+            if (wait.IsCompleted)
+                return new ValueTask<AsyncLockCookie>(Wait(wait.Result));
+            else
```
Comment (grant-d) [inline src/SourceCode.Clay.Threading/AsyncLock.cs]: Remove `else`

File: src/SourceCode.Clay.Threading/SourceCode.Clay.Threading.csproj [hunk 4]
```diff
@@ -8,6 +8,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.9.1" PrivateAssets="All" />
     <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta2-18618-05" PrivateAssets="All" />
+    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.2" />
```
Comment (grant-d) [inline src/SourceCode.Clay.Threading/SourceCode.Clay.Threading.csproj]: Consider multi-targeting netcoreapp2.1/2.2

Comment (jcdickinson) [reply src/SourceCode.Clay.Threading/SourceCode.Clay.Threading.csproj]: System.Threading.Tasks.Extensions is a "null" package for netcore. You only need to multitarget where no polyfill package exists.

**Pull request merged**